spec.html

```diff
@@ -7875,6 +7880,246 @@ <h1>Forward Progress</h1>
+          1. For each WeakMap _map_ such that _map_.[[WeakMapData]] contains a record _r_ such that _r_.[[Key]] is _obj_,
+            1. Remove _r_ from _map_.[[WeakMapData]].
+          1. For each WeakSet _set_ such that _set_.[[WeakSetData]] contains _obj_,
+            1. Remove _obj_ from _set_.[[WeakSetData]].
+      </emu-alg>
+
+      <emu-note>
+        <p>
+          Together with the definition of liveness, this clause prescribes legal
+          optimizations that an implementation may apply regarding WeakRefs.
+        </p>
+
+        <p>
+          It is possible to access an object without observing its
+          identity. Optimizations such as dead variable elimination, and scalar
+          replacement on properties of non-escaping objects whose identity is not
+          observed, are allowed. These optimizations are thus allowed to
+          observably empty WeakRefs that point to such objects.
+        </p>
+
+        <p>
+          On the other hand, if an object's identity is observable, and that
+          object is in the [[WeakRefTarget]] internal slot of a WeakRef, optimizations
+          such as rematerialization that observably empty the WeakRef are
+          prohibited.
+        </p>
+
+        <p>
+          Because calling HostCleanupFinalizationRegistry is optional, registered objects
+          in a FinalizationRegistry do not necessarily hold that FinalizationRegistry live.
+          Implementations may omit FinalizationRegistry callbacks for any reason,
+          e.g., if the FinalizationRegistry itself becomes dead, or if the application
+          is shutting down.
+        </p>
+      </emu-note>
+      <emu-note>
+        <p>
+          Implementations are not obligated to empty WeakRefs for maximal sets
+          of non-live objects.
+        </p>
+        <p>
+          If an implementation chooses a non-live set _S_ in which to empty WeakRefs, it
+          must empty WeakRefs for all objects in _S_ simultaneously. In other
+          words, an implementation must not empty a WeakRef pointing to an object _obj_
+          without emptying out other WeakRefs that, if not emptied, could result
+          in an execution that observes the Object value of _obj_.
+        </p>
+      </emu-note>
+    </emu-clause>
+
+    <emu-clause id="sec-weakref-host-hooks">
+      <h1>Host Hooks</h1>
+
+      <emu-clause id="sec-host-cleanup-finalization-registry" aoid=HostCleanupFinalizationRegistry>
+        <h1>HostCleanupFinalizationRegistry(_finalizationRegistry_)</h1>
+
+        <p>
+          HostCleanupFinalizationRegistry is an implementation-defined abstract
+          operation that is expected to call CleanupFinalizationRegistry(_finalizationRegistry_) at
+          some point in the future, if possible. The host's responsibility is to
+          make this call at a time which does not interrupt synchronous
+          ECMAScript code execution.
```

Comment:
I guess something related to this hook would tie into the incumbent Realm tracking spec work Shu is doing,but this does not need to be in the PR that you send upstream.